Задание 2
Создайте модель под названием model с помощью RandomForestRegressor из модуля
sklearn.ensemble.
Сделайте агрумент n_estimators равным 1000,
max_depth должен быть равен 12 и random_state сделайте равным 42.
Обучите модель на тренировочных данных аналогично тому, как вы обучали модель LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно
применение массивов вместо датафрейма.
Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из
предыдущего задания.
Напишите в комментариях к коду, какая модель в данном случае работает лучше.

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_boston

In [2]:
boston = load_boston()
boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [3]:
data = boston["data"]
feature_names = boston["feature_names"]
target = boston["target"]

In [4]:
X = pd.DataFrame(data, columns=feature_names)
y = pd.DataFrame(target, columns=["price"])

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, shuffle=True, random_state=42)

In [12]:
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(n_estimators=1000, max_depth=12, random_state=42)

In [13]:
model.fit(X_train, y_train.values[:, 0])

RandomForestRegressor(max_depth=12, n_estimators=1000, random_state=42)

In [14]:
y_train_preds = model.predict(X_train)
y_test_preds = model.predict(X_test)

In [15]:
from sklearn.metrics import r2_score

In [16]:
r2_score(y_train, y_train_preds)

0.9784468411299835

In [17]:
r2_score(y_test, y_test_preds)

0.87472606157312

Модель линейной регрессии r2 = 0.71
Модель Случайный лес r2 = 0.87
Модель случайный лес дает гораздо лучший результат, так как при построении каждого следующего дерева учитывает результаты построения предыдущего. А линейная регрессия в принципе не может быть очень точной, так как при линейной зависимости отклонения всегда будут достаточно большими.

*Задание 3
Вызовите документацию для класса RandomForestRegressor,
найдите информацию об атрибуте feature_importances_.
С помощью этого атрибута найдите сумму всех показателей важности,
установите, какие два признака показывают наибольшую важность.

In [27]:
from sklearn.ensemble import RandomForestRegressor

In [28]:
?RandomForestRegressor

feature_importances_ : ndarray of shape (n_features,)
    The impurity-based feature importances.
    The higher, the more important the feature.
    The importance of a feature is computed as the (normalized)
    total reduction of the criterion brought by that feature.  It is also
    known as the Gini importance.

    Warning: impurity-based feature importances can be misleading for
    high cardinality features (many unique values).

In [32]:
feature_importances = pd.DataFrame(zip(*[X_train.columns], model.feature_importances_), 
                                   columns=['feature_name', 'importance'])

feature_importances.sort_values(by='importance', ascending=False, inplace=True)
feature_importances

feature_name  importance
12        LSTAT    0.415847
5            RM    0.402682
7           DIS    0.063973
0          CRIM    0.031676
10      PTRATIO    0.018081
6           AGE    0.014299
4           NOX    0.014269
11            B    0.012451
9           TAX    0.011525
2         INDUS    0.007138
8           RAD    0.005281
1            ZN    0.001543
3          CHAS    0.001236

In [36]:
model.feature_importances_.sum()

1.0

Наибольшую важность имеют признаки LSTAT (0.415847) % lower status of the population и RM (0.402682) average number of rooms per dwelling